### PR TITLE
Chapter 1 examples - start with evaluate implementation

### PR DESCRIPTION
**Pull request by jumarko**
Chapters 1.2 - 1.4 (up to 1.4.3 included).

The most interesting was 1.4.3 implementing sequence of operations via `eprogn`.
Some functions like `invoke`, `lookup`, and `evlis` has to be implemented before their introduction in the book.

### CODE REVIEW

File: clojure/src/ch01_evaluator.clj [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+(atom? {1 2 3 4})
+;; => true
+
+
+(defn evaluate [exp env]
+  (if (atom? exp)
+    ,,,
+    ;; we use `first` instead of `car`
+    (case (first exp)
+      ,,,)))
```
Comment (mrmikardo) [inline clojure/src/ch01_evaluator.clj]: I had no idea that you can sort of iteratively build up function definitions like this? I guess that successive calls to `defn` simply redefine the symbol?

Comment (mrmikardo) [reply clojure/src/ch01_evaluator.clj]: This is cool btw! Makes it super nice to read.

File: clojure/src/ch01_evaluator.clj [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+
+;;; 1.3 (p.4): start building the implementation
+;;; - introduces `atom?` check and `case` for non-atomic forms
+;;; - discusses the difference between 'program' and 'representation';
+;;;   evaluate works on the _representation_ 
+
+;; they define atom as anything that's not a "pair" (cons cell)
+;; - we use `list?` for Clojure which should be good enough
+(def atom? (complement list?))
```
Comment (mrmikardo) [inline clojure/src/ch01_evaluator.clj:15]: I've seen `def` (as opposed to `defn`) used this way before. Is this just a convention, or is there a more in-depth reason for this?

File: clojure/src/ch01_evaluator.clj [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+;; Note: we define our own `wrong` function here
+(defn wrong [msg exp & args]
+  (throw (ex-info msg {:expression exp :rest-args args})))
+
+(defn evaluate [exp env]
+  (if (atom? exp)
+    (cond
+      (symbol? exp) (lookup exp env)
+      ;; TODO: why is vector? here??
+      ;; TODO: what about keywords? nil?
```
Comment (mrmikardo) [inline clojure/src/ch01_evaluator.clj:61]: keywords strike me as autoquotable, too - just return them as the value of `exp`?

File: clojure/src/ch01_evaluator.clj [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      ((some-fn number? string? char? boolean? vector?) exp) exp
+      :else (wrong "Cannot evaluate - unknown atomic expression?" exp))
+
+    ;; we use `first` instead of `car`
+    (case (first exp)
+      ,,,)))
+
+(evaluate 1 {})
+;; => 1
```
Comment (mrmikardo) [inline clojure/src/ch01_evaluator.clj:69]: I'm guessing there's some CiDER variable you can set which automatically writes these to a file?

Comment (mrmikardo) [reply clojure/src/ch01_evaluator.clj:69]: Guessing this? https://emacs.stackexchange.com/a/41660

File: clojure/src/demo.clj [hunk 5]
```diff
@@ -0,0 +1,22 @@
+(ns demo
```
Comment (mrmikardo) [inline clojure/src/demo.clj:1]: Thanks for posting this here too ;)

**Review by mrmikardo — COMMENTED** [review]
Asked some newbie Clojure questions!

File: clojure/src/ch01_evaluator.clj (hunk 1 above)
Comment (jumarko) [reply clojure/src/ch01_evaluator.clj]: Yes, exactly - it's typically an error to have duplicate definitions but for a nice "trick" to use for building up the stuff incrementally.

File: clojure/src/ch01_evaluator.clj (hunk 2 above)
Comment (jumarko) [reply clojure/src/ch01_evaluator.clj:15]: `defn` is a macro built on top of `def` and `fn`.

But here [`complement`](https://clojuredocs.org/clojure.core/complement) is a higher-order function that already a function so we just use `def`.
It's a sort of _point-free_ style of programming which I don't really like so use it very sparingly :).
An alternative would be: 
```
(defn atom? [exp]
  (not (list? exp)))
```

File: clojure/src/ch01_evaluator.clj (hunk 3 above)
Comment (jumarko) [reply clojure/src/ch01_evaluator.clj:61]: Exactly, that's what I should do. I just left it out temporarily so it's the first thing we can do tomorrow.

File: clojure/src/ch01_evaluator.clj (hunk 4 above)
Comment (jumarko) [reply clojure/src/ch01_evaluator.clj:69]: I'm using the command I showed you yesterday: `cider-eval-defun-to-comment` (there's also `cider-pprint-eval-defun-to-comment`).

**Pull request merged**